raysort/main_push.py

```diff
@@ -201,19 +206,73 @@ def final_merge(
     cfg: AppConfig,
     worker_id: PartId,
     reduce_idx: PartId,
-    *parts: list[np.ndarray],
-) -> PartInfo:
+    parts: list[np.ndarray],
+    subpart_idx: PartId = 1,
+    level: int = 1,
+) -> list[PartInfo]:
     logging_utils.init()
+
+    def id_print(*output):
+        print(":) final_merge identifier:", str(worker_id) + "_" + str(reduce_idx), "(" + str(subpart_idx) + ")", *output)
+
+    id_print("| level =", level)
+
     with tracing_utils.timeit("reduce"):
         M = len(parts)
+        if M == 0:
+            return []
+
+        ray.wait(parts, num_returns=len(parts))
+
+        part_locs = ray.experimental.get_object_locations(parts)
+        part_sizes = [loc["object_size"] for loc in part_locs.values()]
+        
+        id_print("all part_sizes:", part_sizes)
+
+        parts_memory_gb = sum(part_sizes) / 1_000_000_000
+        id_print("parts_memory_gb:", parts_memory_gb)
+        # if M > 1 and parts_memory_gb > cfg.dynamic_repartition_threshold_gb:
+        THRESHOLD_GB = 8
+        if M > 1 and parts_memory_gb > THRESHOLD_GB:
+            id_print("exceeded threshold, recursing")
+            # split parts
+            part_0 = ray.get(parts[0])
+            pivot = sort_utils.get_median_key(part_0)
+            separated_parts = []
+            for part in parts:
+                separated_parts.append(sort_utils.split_part.remote(part, pivot, (str(worker_id) + "_" + str(reduce_idx))))
+            
+            first_parts = [a for a,_ in separated_parts if a]
+            second_parts = [b for _,b in separated_parts if b]
+
+            first_half = final_merge.options(**ray_utils.current_node_aff()).remote(
+                    cfg, worker_id, reduce_idx, first_parts, subpart_idx = subpart_idx * 2, level = level + 1
+                ) or []
+            second_half = final_merge.options(**ray_utils.current_node_aff()).remote(
+                    cfg, worker_id, reduce_idx, second_parts, subpart_idx = subpart_idx * 2 + 1, level = level + 1
+                ) or []
+            
+            grouped = [first_half, second_half]
+            ray_get = ray.get(grouped)
+
+            output = []
```

Comment:
Can use list.flatten: https://stackoverflow.com/questions/952914/how-do-i-make-a-flat-list-out-of-a-list-of-lists